File: bush-maths/python/src/pools/weighted/weighted_data.py
from dataclasses import dataclass, fields
from typing import List, Optional

from src.common.base_pool_state import BasePoolState


@dataclass
class WeightedState(BasePoolState):
    weights: List[int]
    # Per-token minimum live balance (scaled 18) enforced by WeightedPool: max(1e6, 10^(18 - decimals)).
    # Read from `getMinTokenBalances()`; defaults from the scaling factors when omitted.
    min_token_balances: Optional[List[int]] = None

    def __init__(self, **kwargs):
        # Set pool_type before initializing BasePoolState
        kwargs["pool_type"] = "WEIGHTED"

        # Get BasePoolState fields
        base_fields = {f.name for f in fields(BasePoolState)}
        base_kwargs = {k: v for k, v in kwargs.items() if k in base_fields}

        # Initialize BasePoolState with its fields
        super().__init__(**base_kwargs)

        # Get WeightedState fields
        custom_fields = {f.name for f in fields(WeightedState)}

        # Set WeightedState fields
        for field in custom_fields:
            if field in kwargs:
                setattr(self, field, kwargs[field])


def map_weighted_state(pool_state: dict) -> WeightedState:
    return WeightedState(
        # BasePoolState fields
        pool_address=pool_state["poolAddress"],
        tokens=pool_state["tokens"],
        scaling_factors=pool_state["scalingFactors"],
        token_rates=pool_state["tokenRates"],
        balances_live_scaled18=pool_state["balancesLiveScaled18"],
        swap_fee=pool_state["swapFee"],
        aggregate_swap_fee=pool_state["aggregateSwapFee"],
        total_supply=pool_state["totalSupply"],
        supports_unbalanced_liquidity=pool_state.get(
            "supportsUnbalancedLiquidity", True
        ),
        hook_type=pool_state.get("hookType", None),
        # WeightedState fields
        weights=pool_state["weights"],
        min_token_balances=pool_state.get("minTokenBalances"),
    )

File: maths/rust/src/pools/weighted/weighted_data.rs
//! Weighted pool data structures

use crate::common::types::BasePoolState;
use alloy_primitives::U256;
use serde::{Deserialize, Serialize};

/// Weighted pool state (extends BasePoolState with weighted-specific fields)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WeightedState {
    /// Base pool state
    #[serde(flatten)]
    pub base: BasePoolState,
    /// Normalized weights (scaled 18)
    pub weights: Vec<U256>,
    /// Per-token minimum live balance (scaled 18) enforced by WeightedPool: max(1e6, 10^(18 - decimals)).
    /// Read from `getMinTokenBalances()`; derived from the scaling factors when omitted.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub min_token_balances: Option<Vec<U256>>,
}

impl WeightedState {
    /// Create a new weighted pool state
    pub fn new(base: BasePoolState, weights: Vec<U256>) -> Self {
        Self {
            base,
            weights,
            min_token_balances: None,
        }
    }

    /// Get the base pool state
    pub fn base(&self) -> &BasePoolState {
        &self.base
    }

    /// Get the weights
    pub fn weights(&self) -> &[U256] {
        &self.weights
    }
}

impl From<WeightedState> for BasePoolState {
    fn from(weighted_state: WeightedState) -> Self {
        weighted_state.base
    }
}

impl AsRef<BasePoolState> for WeightedState {
    fn as_ref(&self) -> &BasePoolState {
        &self.base
    }
}

impl From<WeightedState> for crate::common::types::PoolState {
    fn from(state: WeightedState) -> Self {
        crate::common::types::PoolState::Weighted(state)
    }
}

impl WeightedState {
    /// From the `pool` block of a test-data file (see `crate::common::json`).
    pub fn from_json(pool: &crate::common::json::PoolJson, base: BasePoolState) -> Result<Self, String> {
        Ok(WeightedState {
            base,
            weights: pool.arr("weights")?,
            min_token_balances: pool.arr("minTokenBalances").ok(),
        })
    }
}
